crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -475,34 +475,60 @@ async fn cache_latest_events(
-                    // `Store` has an async API, we have no choice than doing this manually.
-                    //
-                    // First off, look up the sender's profile from the `changes`.
+                    // In order to make the latest event fast to read, we want to keep the
+                    // associated sender in cache. This is a best-effort to gather enough
+                    // information for creating a user profile as fast as possible. If information
+                    // are missing, let's go back on the “slow” path.
+
                     let mut sender_profile = None;
+                    let mut sender_name_is_ambiguous = None;
 
+                    // First off, look up the sender's profile from the `StateChanges`, they are
+                    // likely to be the most recent information.
                     if let Some(changes) = changes {
                         sender_profile = changes
                             .profiles
                             .get(room.room_id())
-                            .map(|profiles_by_user| profiles_by_user.get(timeline_event.sender()))
-                            .flatten()
+                            .and_then(|profiles_by_user| {
+                                profiles_by_user.get(timeline_event.sender())
+                            })
                             .cloned();
+
+                        if let Some(sender_profile) = sender_profile.as_ref() {
+                            sender_name_is_ambiguous = sender_profile
+                                .as_original()
+                                .and_then(|profile| profile.content.displayname.as_ref())
+                                .and_then(|display_name| {
+                                    changes.ambiguity_maps.get(room.room_id()).and_then(
+                                        |map_for_room| {
+                                            map_for_room
+                                                .get(display_name)
+                                                .map(|user_ids| user_ids.len() > 1)
+                                        },
+                                    )
+                                });
+                        }
                     }
 
-                    // Second, look up the sender's profile from the storage if none was found.
+                    // Otherwise, look up the sender's profile from the `Store`.
                     if sender_profile.is_none() {
                         if let Some(store) = store {
                             sender_profile = store
                                 .get_profile(room.room_id(), timeline_event.sender())
                                 .await
                                 .ok()
                                 .flatten();
+
+                            // TODO: need to update `sender_name_is_ambiguous`,
+                            // but how?
```

Comment:
Best effort would be to look at the `changes.ambiguity_map` here too (and refactor to code to avoid the above to be duplicated here). The changes could be there while the sender profile might not be in changes.

Reply:
I can't look inside `changes` at this step because it's supposed to be `None`. Let's keep the `TODO` for now, and let's see if it creates some issues.